pkg/controller/attach_detach.go

```diff
@@ -599,128 +585,130 @@ func (r *ReconcileAttachDetach) recreateAzVolumeAttachment(ctx context.Context,
+			azVolumeAttachmentName := azureutils.GetAzVolumeAttachmentName(volumeName, nodeName)
 			r.azVolumeAttachmentToVaMap.Store(azVolumeAttachmentName, volumeAttachment.Name)
 
-			desiredAzVolumeAttachment := &azdiskv1beta2.AzVolumeAttachment{
-				ObjectMeta: metav1.ObjectMeta{
-					Name: azVolumeAttachmentName,
-					Labels: map[string]string{
-						consts.NodeNameLabel:   nodeName,
-						consts.VolumeNameLabel: volumeName,
-					},
-					// if the volumeAttachment shows not yet attached, and VolumeAttachRequestAnnotation needs to be set from the controllerserver
-					// if the volumeAttachment shows attached but the actual volume isn't due to controller restart, VolumeAttachRequestAnnotation needs to be set by the noderserver during NodeStageVolume
-					Finalizers: []string{consts.AzVolumeAttachmentFinalizer},
-				},
-				Spec: azdiskv1beta2.AzVolumeAttachmentSpec{
-					VolumeName:    volumeName,
-					VolumeID:      pv.Spec.CSI.VolumeHandle,
-					NodeName:      nodeName,
-					RequestedRole: azdiskv1beta2.PrimaryRole,
-					VolumeContext: pv.Spec.CSI.VolumeAttributes,
-				},
-			}
-			azureutils.AnnotateAPIVersion(desiredAzVolumeAttachment)
+			desiredAzVolumeAttachment := r.generateDesiredAzVolumeAttachment(azVolumeAttachmentName, nodeName, volumeName, volumeHandle, azdiskv1beta2.PrimaryRole, volumeAttributes[volumeName])
 
-			statusUpdateRequired := true
-			// check if the CRI exists already
-			azVolumeAttachment, err := azureutils.GetAzVolumeAttachment(ctx, r.cachedClient, r.azClient, azVolumeAttachmentName, r.config.ObjectNamespace, false)
+			err = r.recreateAzVolumeAttachment(ctx, azVolumeAttachmentName, desiredAzVolumeAttachment)
 			if err != nil {
-				if errors.IsNotFound(err) {
-					w.Logger().Infof("Recreating AzVolumeAttachment(%s)", azVolumeAttachmentName)
-
-					azVolumeAttachment, err = r.azClient.DiskV1beta2().AzVolumeAttachments(r.config.ObjectNamespace).Create(ctx, desiredAzVolumeAttachment, metav1.CreateOptions{})
-					if err != nil {
-						w.Logger().Errorf(err, "failed to create AzVolumeAttachment (%s) for volume (%s) and node (%s): %v", azVolumeAttachmentName, *pvName, nodeName, err)
-						return syncedVolumeAttachments, volumesToSync, err
-					}
-				} else {
-					w.Logger().Errorf(err, "failed to get AzVolumeAttachment (%s): %v", azVolumeAttachmentName, err)
-					return syncedVolumeAttachments, volumesToSync, err
-				}
-			} else if apiVersion, ok := azureutils.GetFromMap(azVolumeAttachment.Annotations, consts.APIVersion); !ok || apiVersion != azdiskv1beta2.APIVersion {
-				w.Logger().Infof("Found AzVolumeAttachment (%s) with older api version. Converting to apiVersion(%s)", azVolumeAttachmentName, azdiskv1beta2.APIVersion)
-
-				for k, v := range desiredAzVolumeAttachment.Labels {
-					azVolumeAttachment.Labels = azureutils.AddToMap(azVolumeAttachment.Labels, k, v)
-				}
+				return syncedVolumeAttachments, volumeHandles, volumeAttributes, err
```

Comment:
If there is an error processing one volume, we should log the error and continue rather than terminate the loop.

Reply:
Done.